// reg-publish-s3-plugin/lib/s3-bucket-preparer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketPreparer = void 0;
class S3BucketPreparer {
    inquire() {
        return [
            {
                name: "bucketName",
                type: "input",
                message: "Bucket name",
            },
        ];
    }
    prepare(config) {
        return __awaiter(this, void 0, void 0, function* () {
            this._logger = config.logger;
            const { bucketName } = config.options;
            const pluginConfig = {
                bucketName,
            };
            return pluginConfig;
        });
    }
}
exports.S3BucketPreparer = S3BucketPreparer;
